activerecord/lib/active_record/relation/predicate_builder.rb

```diff
@@ -137,7 +138,11 @@ def convert_dot_notation_to_hash(attributes)
     end
 
     def handler_for(object)
-      @handlers.detect { |klass, _| klass === object }.last
+      if handler = @handlers_cache[object.class]
```

Comment:
@sgrif Any ideas how we should cache this properly? I'm afraid that the cache might be vulnerable because I can pass  in random objects and the cache will just keep growing.